### PR TITLE
Remove case insensitivity in data types mapping 

### PR DESCRIPTION
**Pull request by Ivanidzo4ka**
## Description
https://github.com/microsoft/fhir-server/blob/2790a877ec4397f9c61c129a99dbf9e6145ad84a/src/Microsoft.Health.Fhir.Core/Features/Search/TypedElementSearchIndexer.cs#L190

`.Select(fhirPathExpression)` is case sensitive so there is different results in doing `value.as(DateTime)` and `value.as(dateTime)`. First yield nothing, and second produce dateTime node.

This change normalizes case in search-parameter json files and put restriction in place in order to avoid this situation in future.

On a side note: all these search parameters currently produce nothing and can't be searched, but they are in `searchable` state and I'm not sure what we can do to reindex them, or put them in `supported` state.

## Related issues
Addresses [#2020].

## Testing
Describe how this change was tested.

## FHIR Team Checklist
- [ ] **Update the title** of the PR to be succinct and less than 50 characters
- [ ] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [ ] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [ ] Tag the PR with **Azure API for FHIR** if this will release to the managed service
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

**Comment by rotodd** [comment]
To have it handy, here are the search parameters that appear to be affected:

R4:

`http://hl7.org/fhir/SearchParameter/Observation-code-value-date`

R5:

`http://hl7.org/fhir/SearchParameter/Observation-code-value-date`

Stu3:
```
http://hl7.org/fhir/SearchParameter/clinical-date
http://hl7.org/fhir/SearchParameter/CommunicationRequest-occurrence
http://hl7.org/fhir/SearchParameter/ConceptMap-source-uri
http://hl7.org/fhir/SearchParameter/ConceptMap-target-uri
http://hl7.org/fhir/SearchParameter/DeviceRequest-event-date
http://hl7.org/fhir/SearchParameter/Goal-start-date
http://hl7.org/fhir/SearchParameter/Goal-target-date
http://hl7.org/fhir/SearchParameter/Observation-code-value-date
http://hl7.org/fhir/SearchParameter/Observation-value-date
http://hl7.org/fhir/SearchParameter/Observation-value-string
http://hl7.org/fhir/SearchParameter/Patient-death-date
```

**Comment by brendankowitz** [comment]
Adding a reference to #901

File: src/Microsoft.Health.Fhir.Core/Data/R4/search-parameters.json [hunk 1]
```diff
@@ -45840,7 +45840,7 @@
       },
       {
         "definition" : "http://hl7.org/fhir/SearchParameter/Observation-value-date",
-        "expression" : "value.as(DateTime) | value.as(Period)"
+        "expression" : "value.as(dateTime) | value.as(Period)"
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Data/R4/search-parameters.json:45843]: This sounds like a parser bug too. `If the identifier cannot be resolved to a valid type identifier, the evaluator will throw an error.` http://hl7.org/fhirpath/N1/#as-type-specifier

**Review by brendankowitz — DISMISSED** [review]
I don't think we should edit the search-parameters.json file.

**Comment by brendankowitz** [comment]
@feordin @namadabu If these expressions change, will that give us a different hash and allow reindexing? Would there be any other side-effects here?

**Comment by feordin** [comment]
@brendankowitz Yes, the expression value is used when computing the hash, so it would change the hash value.  However, as these parameters are part of the "out of the box" parameters, the startup code by default marks all of those parameters as fully indexed, a status of "enabled".  The Reindex logic will not detect them as needing to be indexed.  We would need an additional code changes to detect changes in the search-parameters.json file.

**Comment by rbans96** [comment]
We should redo this for new parts of the spec(R4B, R5). @brendankowitz @feordin will create a user story for this.

**Pull request closed without merging**